torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -0,0 +1,208 @@
+//                 +------+
+// +------+------+ +------+
+// |      |      | |      |
+// |  L1  |  L2  | |  O   |
+// |      |      | |      |
+// +------+------+ +------+
+
+
+
+// Note [Further optimizations]
+// It would be straightforward to extend the TreeToken class to also detect if all
+// MMs had the same lhs/rhs. In such case it's more efficient to expand the lhs
+// and use bmm + sum instead of repeating it in memory via concat.
+
+// Note [Overlapping trees]
+// Additionally it wouldn't be too hard to add support for partially overlapping
+// trees. Right now the it's forbidden in the algorithm (only a single tree will
+// be allowed), so theoretically we might miss some optimization options, especially
+// that the rejected tree could be much larger. I didn't implement that because it's
+// not necessary for the simple RNN cases I saw, so I decided to keep stuff simple.
+// If we ever get around implementing this, the right solution is probably to fuse
+// MMs for the common part, and assume it's an input leaf for the outer two parts
+// (I don't think it's beneficial to recompute, unless the subtree is super small,
+// but let's not get into such details).
+
+// The algorithm we're using is simple. We're iterating through the graph in the
+// topological order and labeling nodes with TreeTokens. Then, we look for roots of
+// the trees we formed and fuse them.
+
+enum class Side {
+  LHS,
+  RHS
+};
+
+// Tunable parameter. Set to something larger if it turns out to be better.
+static constexpr std::size_t min_fusion_size = 2;
+
+static std::array<int64_t, 2> as_array(at::IntList sizes) {
+  JIT_ASSERT(sizes.size() == 2);
+  std::array<int64_t, 2> arr;
+  arr[0] = sizes[0];
+  arr[1] = sizes[1];
+  return arr;
+}
+
+// TreeTokens will be used to label nodes of the graph, if the nodes will fit
+// our mm/add tree pattern. Basically we do dynamic programming on DAGs, where
+// when we reach node N with inputs A and B, then A and B have already been
+// procesed, and we can try to unify their TreeTokens (if they have them)
+// and build a larger tree.
+struct TreeToken {
+  uint64_t tree_size = 0; // NOTE: measured in number of leaves i.e. mm ops
+  std::array<int64_t, 2> lhs_sizes;
+  std::array<int64_t, 2> rhs_sizes;
+  Node *node = nullptr;
+  bool is_root = false;
+
+  static TreeToken fromMM(Node *mm) {
+    TreeToken token;
+    token.tree_size = 1;
+    Value *lhs = mm->inputs()[0];
+    Value *rhs = mm->inputs()[1];
+    token.lhs_sizes = as_array(lhs->type()->expect<TensorType>()->sizes());
+    token.rhs_sizes = as_array(rhs->type()->expect<TensorType>()->sizes());
+    token.node = mm;
+    token.is_root = true;
+    return token;
+  }
+
+  static TreeToken unify(Node *add, TreeToken& l, TreeToken& r) {
+    TreeToken token;
+    // See Note [Overlapping trees]
+    if (&l == &r || !l.is_root || !r.is_root)
+      return token;
+    // We can batch the tree only if all sizes match, because we need to
```

Comment:
What is a case when the sizes will not match? Is it for rare (1xN) + (M x N) style cases?

Reply:
Yeah I realized later today that since our IR adds are non-broadcasting then we can't have mismatched shapes. Still, the checks are cheap, and it's kind of convenient to keep this information around, but I'm ok with removing them if you think that's better

Reply:
Keep them in, that way if we do allow broadcasting later we don't have to change this.